R/drive_ls.R

```diff
@@ -26,51 +28,161 @@
 #' }
 #'
 #' @export
-drive_ls <- function(search = NULL, ..., fixed = FALSE, verbose = TRUE){
+drive_ls <- function(path = NULL,
+                     pattern = NULL,
+                     ...,
+                     fixed = FALSE,
+                     verbose = TRUE){
+
+  folder <- NULL
 
-  request <- build_drive_ls(...)
+  if (!is.null(path)){
+    folder <- climb_folders(path = path)
+  }
+
+  request <- build_drive_ls(..., folder = folder)
   response <- make_request(request)
-  process_drive_ls(response = response, search = search, fixed = fixed, verbose = verbose)
+  process_drive_ls(
+    response = response,
+    pattern = pattern,
+    fixed = fixed,
+    verbose = verbose
+  )
+
+}
+
+climb_folders <- function(path = NULL, my_folders = drive_folders()){
+
+  folders <- unlist(strsplit(path, "/"))
+  my_folder_ids <- purrr::map(folders, folder_ids, folder_tbl = my_folders)
+
+  if (!any(my_folder_ids[[1]]$root)){
+    spf("We could not find a folder named '%s' in your 'My Drive' (root) directory.", folders[1])
+  }
 
+  #it is v silly that Google Drive allows this but...
+  if (!sum(my_folder_ids[[1]]$root) == 1){
+    spf("It seems you have more than one folder named '%s' in your 'My Drive' (root) directory.", folders[1])
+  }
+
+  if (length(my_folder_ids) == 1){
+    return(my_folder_ids[[1]]$id[my_folder_ids[[1]]$root == TRUE])
+  }
+  len <- length(my_folder_ids)
+
+  while (nrow(my_folder_ids[[len - 1]]) != 1){
```

Comment:
yes - I think I've got something similar to that in my [newest iteration](https://github.com/tidyverse/googledrive/blob/drive_list/R/drive_list.R#L49-L58) - one tricky part is that you may have x folders named `foo` and y named `bar` etc. so it isn't always checking if it's in i-1, it could be that you have something like
```
name  location
foo      root
foo      bar
bar      foo
baz     bar
```
but I *think* I've solved that by [only setting the `parent_id` if it is correct](https://github.com/tidyverse/googledrive/blob/drive_list/R/drive_list.R#L53-L56). I would like to refactor this so it isn't a for loop (which I'm sure is very doable) but I am a bit sleepy. Thank you for walking through this with me I feel like I've got you really getting into the weeds on this one 🌾